### PR TITLE
Registration

### PR DESCRIPTION
**Pull request by 3NFBAGDU**
This change is [Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)

### CODE REVIEW

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 1 of 35 files reviewed, 1 unresolved discussion (waiting on @3NFBAGDU and @bakurits)

___
*[.gitignore, line 3 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGD4Uq77Q5wf3Q-LcW:-MMGD4Us9wyVfalhkIek:b-w6fzg2) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/.gitignore#L3)):*
> ```ignore
> credentials
> .idea
> cmd/webapp/config.json
> ```

Do we need both of these?

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 4 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 5 of 35 files reviewed, 2 unresolved discussions (waiting on @3NFBAGDU and @bakurits)

___
*[cmd/fileshare/state.json, line 1 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGDjkE8dHllljRSTFU:-MMGDjkE8dHllljRSTFV:b7slcur) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/cmd/fileshare/state.json#L1)):*
> ```JSON
> {"credentialsPath":"../credentials"}
> ```

this should be removed

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 2 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 7 of 35 files reviewed, 3 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cfg/config.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGETsAD_UH59bfNG6I:-MMGETsB1qyOA67zK0JZ:b-71avg5) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cfg/config.go#L7)):*
> ```Go
> 	"github.com/jinzhu/configor"
> 	"github.com/pkg/errors"
> 	"io/ioutil"
> ```

This should be moved up

**Review by bakurits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 7 of 35 files reviewed, 4 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cfg/config.go, line 1 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGEbNc6evW-NejwnU8:-MMGEbNc6evW-NejwnU9:b-bdbf2d) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cfg/config.go#L1)):*
> ```Go
> package cfg
> ```

Documentation missing

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 4 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 11 of 35 files reviewed, 5 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cmd/drivemanager/list.go, line 5 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGEpyu90YSTAHWQJee:-MMGEpyu90YSTAHWQJef:b-zgtbpn) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cmd/drivemanager/list.go#L5)):*
> ```Go
> import (
> 	"fmt"
> 	"github.com/bakurits/fileshare/pkg/drive"
> ```

There should be extra line

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 2 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 13 of 35 files reviewed, 7 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cmd/drivemanager/uploadfile.go, line 5 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGEvzU0nTvbwUiePfM:-MMGEvzU0nTvbwUiePfN:b-zgtbpn) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cmd/drivemanager/uploadfile.go#L5)):*
> ```Go
> import (
> 	"github.com/bakurits/fileshare/pkg/drive"
> 	"github.com/pkg/errors"
> ```

There should be extra line
___
*[pkg/cmd/drivemanager/util.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGEzGy-7m9v8nT5kOD:-MMGEzGy-7m9v8nT5kOE:b7lfrzk) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cmd/drivemanager/util.go#L7)):*
> ```Go
> 	"github.com/bakurits/fileshare/pkg/cfg"
> 	"github.com/pkg/errors"
> 	"os"
> ```

This should be above

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 14 of 35 files reviewed, 8 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cmd/mail/sendattachment.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGF6Vs59r3_vlk8r15:-MMGF6Vs59r3_vlk8r16:bgbs51o) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cmd/mail/sendattachment.go#L7)):*
> ```Go
> 	"github.com/pkg/errors"
> 	"github.com/spf13/cobra"
> 	"path/filepath"
> ```

There should be above

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 2 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 16 of 35 files reviewed, 9 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cmd/mail/util.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGFBCR29tT3Wpu6XkP:-MMGFBCR29tT3Wpu6XkQ:b7lfrzk) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cmd/mail/util.go#L7)):*
> ```Go
> 	"github.com/bakurits/fileshare/pkg/cfg"
> 	"github.com/pkg/errors"
> 	"os"
> ```

This should be above

**Review by bakurits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 16 of 35 files reviewed, 10 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/drive/download_test.go, line 6 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGFJLu6NtHzlyxaL59:-MMGFJLw-Mtg7SYRnkSL:b7lfrzk) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/drive/download_test.go#L6)):*
> ```Go
> 	"github.com/bakurits/fileshare/pkg/auth"
> 	"github.com/stretchr/testify/assert"
> 	"testing"
> ```

This should be above

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 2 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 18 of 35 files reviewed, 11 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/drive/get_test.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGFWGQEZrdpnfeOup3:-MMGFWGQEZrdpnfeOup4:b7lfrzk) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/drive/get_test.go#L7)):*
> ```Go
> 	"github.com/bakurits/fileshare/pkg/auth"
> 	"github.com/stretchr/testify/assert"
> 	"testing"
> ```

This should be above

File: pkg/drive/testutils.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package drive
+
+import "github.com/bakurits/fileshare/pkg/cfg"
+
+func getTestConfig() (cfg.Config, error) {
+	conf := cfg.Config{
+		GoogleCredentialsPath: "/home/bakurits/Documents/Programming/Utilities/fileshare/credentials/web_app_credentials.json",
```
Comment (3NFBAGDU) [inline pkg/drive/testutils.go:7]: maybe test config should be created.

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 19 of 35 files reviewed, 12 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/drive/list_test.go, line 22 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGFYpe9p1PQnCpGKbi:-MMGFYpe9p1PQnCpGKbj:bebaq26) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/drive/list_test.go#L22)):*
> ```Go
> 	srv, err := NewService(client.Client)
> 	is.NoError(err)
> 	files := srv
> ```

Output should be file names

**Review by bakurits — COMMENTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 20 of 35 files reviewed, 12 unresolved discussions (waiting on @3NFBAGDU)

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 2 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 22 of 35 files reviewed, 13 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/webapp/db/models.go, line 9 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGFqn4CHc_UzazBSfM:-MMGFqn4CHc_UzazBSfN:b7lfrzk) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/webapp/db/models.go#L9)):*
> ```Go
> 	"github.com/pkg/errors"
> 	"golang.org/x/oauth2"
> 	"time"
> ```

This should be above

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 23 of 35 files reviewed, 14 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/webapp/db/repository.go, line 5 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGFuOs19ONlyjSzEcR:-MMGFuOs19ONlyjSzEcS:b7lfrzk) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/webapp/db/repository.go#L5)):*
> ```Go
> import (
> 	"gorm.io/driver/sqlite"
> 	"log"
> ```

This should be above

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 3 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 26 of 35 files reviewed, 15 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/webapp/server/util.go, line 27 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGG9da1-5fH9mDjTe1:-MMGG9da1-5fH9mDjTe2:bht276k) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/webapp/server/util.go#L27)):*
> ```Go
> 	return func(c *gin.Context) {
> 		session := sessions.Default(c)
> 		email := session.Get("email")
> ```

Session keys must be stored in constant variables

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 27 of 35 files reviewed, 16 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/webapp/server/web.go, line 82 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGGYR66AOIPZTkons6:-MMGGYR70HNlbciCzDAf:bya048u) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/webapp/server/web.go#L82)):*
> ```Go
> 		}
> 		state := randToken()
> 		session.Set("state", state)
> ```

Session keys must be constant variables

**Review by bakurits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 27 of 35 files reviewed, 18 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/webapp/server/web.go, line 156 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGGiDB37UxQKZt0AEj:-MMGGiDB37UxQKZt0AEk:b-ck6g8t) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/webapp/server/web.go#L156)):*
> ```Go
> }
> 
> func (s *Server) getEmailFromPasswordRecoveryRequest(c *gin.Context) string {
> ```

This should be in utils

**Review by bakurits — CHANGES_REQUESTED** [review]
> Reviewed 7 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 34 of 35 files reviewed, 19 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/auth/auth.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGGwui3WvVE5mtFeQU:-MMGGwujBFMgGyxOXoJa:b7m1f5w) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/auth/auth.go#L7)):*
> ```Go
> 	"encoding/json"
> 	"fmt"
> 	"github.com/bakurits/fileshare/pkg/cfg"
> ```

This should be below

**Review by bakurits — COMMENTED** [review]
> Reviewed 1 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 19 unresolved discussions (waiting on @3NFBAGDU)

**Comment by 3NFBAGDU** [comment]
> > Reviewed 1 of 35 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)_ status: 1 of 35 files reviewed, 1 unresolved discussion (waiting on @3NFBAGDU and @bakurits)
> 
> _[.gitignore, line 3 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGD4Uq77Q5wf3Q-LcW:-MMGD4Us9wyVfalhkIek:b-w6fzg2) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/.gitignore#L3)):_
> 
> > ```gitignore
> > credentials
> > .idea
> > cmd/webapp/config.json
> > ```
> 
> Do we need both of these?

**Review by bakurits — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 21 of 35 files reviewed, 17 unresolved discussions (waiting on @3NFBAGDU and @bakurits)

___
*[pkg/drive/testutils.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGBwIWJ3HzOnS2WnbQ-r1-7:-MMGSeWA40ddJt7rYbJ-:b9rhcba) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/drive/testutils.go#L7)):*
<details><summary><i>Previously, 3NFBAGDU (Giorgi Baghdavadze) wrote…</i></summary><blockquote>

maybe test config should be created.
</blockquote></details>

It's good idea

**Review by bakurits — COMMENTED** [review]
> Reviewed 7 of 11 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 34 of 35 files reviewed, 10 unresolved discussions (waiting on @3NFBAGDU and @bakurits)

**Review by bakurits — COMMENTED** [review]
> Reviewed 1 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 34 of 35 files reviewed, 9 unresolved discussions (waiting on @3NFBAGDU and @bakurits)

**Review by bakurits — COMMENTED** [review]
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 5 unresolved discussions (waiting on @3NFBAGDU)

**Review by bakurits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 6 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/auth/auth.go, line 3 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGVHo29UeP5v5WR9Nj:-MMGVHo29UeP5v5WR9Nk:b-ckb5ov) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/auth/auth.go#L3)):*
> ```Go
> package auth
> 
> import (
> ```

sort imports

**Review by bakurits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 7 unresolved discussions (waiting on @3NFBAGDU)

___
*[pkg/cmd/mail/util.go, line 7 at r1](https://reviewable.io/reviews/bakurits/fileshare/9#-MMGVS6l1my3m493gwDp:-MMGVS6l1my3m493gwDq:b-ckb5ov) ([raw file](https://github.com/bakurits/fileshare/blob/474e51a68934e0141fd9f147c83981bb3c98b0fc/pkg/cmd/mail/util.go#L7)):*
> ```Go
> 	"github.com/bakurits/fileshare/pkg/cfg"
> 	"github.com/pkg/errors"
> 	"os"
> ```

sort imports

**Review by bakurits — COMMENTED** [review]
> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 4 unresolved discussions (waiting on @3NFBAGDU)

**Review by bakurits — COMMENTED** [review]
> Reviewed 1 of 18 files at r7, 2 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: 24 of 36 files reviewed, 2 unresolved discussions (waiting on @3NFBAGDU and @bakurits)

**Review by bakurits — COMMENTED** [review]
> Reviewed 11 of 18 files at r7, 1 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 2 unresolved discussions (waiting on @3NFBAGDU)

**Review by bakurits — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/bakurits/fileshare/9)* status: all files reviewed, 1 unresolved discussion (waiting on @3NFBAGDU)

**Pull request merged**